base/lock.jl

```diff
@@ -314,6 +314,36 @@ function acquire(s::Semaphore)
     return
 end
 
+"""
+    acquire(f::Function, s::Semaphore)
+
+Execute `f` after acquiring from Semaphore `s`,
+and `release` on completion or error.
+
+For example, a do-block form that ensures only 2
+calls of `foo` will be active at the same time:
+
+```julia
+s = Base.Semaphore(2)
+@sync for _ in 1:100
+    Threads.@spawn begin
+        Base.acquire(s) do
+            foo()
+        end
+    end
+end
+```
+"""
+function acquire(f::Function, s::Semaphore)
```

Comment:
I suggest avoiding `::Function` in any new API design. See #43491 for a relevant discussion.

```suggestion
function acquire(f, s::Semaphore)
```

But it's a bit tricky for `acquire` since it's reasonable to have all of

1. `acquire(f, ::Semaphore)`
2. `acquire(f, ::Semaphore, ncounts::Integer)`: acquire `ncounts` resources
3. `acquire(::Semaphore, ncounts::Integer)`
4. `acquire(::Semaphore)`

and it makes it impossible to infer the semantics of `acquire(a, b)` just from the syntax (maybe it's OK)

Probably we can worry about the generic count version later since it needs to pass "token" to `release` for propagating released count anyway. Maybe we can call it `token = acquiremany(semaphore, ncounts)`.